proxy.go

```diff
@@ -0,0 +1,288 @@
+	RequestCh            chan<- invocation.Invocation
+	SerializationService iserialization.SerializationService
+	PartitionService     *cluster.PartitionService
+	UserEventDispatcher  *event.DispatchService
+	ClusterService       *cluster.ServiceImpl
+	InvocationFactory    *cluster.ConnectionInvocationFactory
+	ListenerBinder       *cluster.ConnectionListenerBinderImpl
+	SmartRouting         bool
+	Logger               ilogger.Logger
+}
+
+func (b creationBundle) Check() {
+	if b.RequestCh == nil {
+		panic("RequestCh is nil")
+	}
+	if b.SerializationService == nil {
+		panic("SerializationService is nil")
+	}
+	if b.PartitionService == nil {
+		panic("PartitionService is nil")
+	}
+	if b.UserEventDispatcher == nil {
+		panic("UserEventDispatcher is nil")
+	}
+	if b.ClusterService == nil {
+		panic("ClusterService is nil")
+	}
+	if b.InvocationFactory == nil {
+		panic("ConnectionInvocationFactory is nil")
+	}
+	if b.ListenerBinder == nil {
+		panic("ListenerBinder is nil")
+	}
+	if b.Logger == nil {
+		panic("Logger is nil")
+	}
+}
+
+type proxy struct {
+	requestCh            chan<- invocation.Invocation
+	serializationService iserialization.SerializationService
+	partitionService     *cluster.PartitionService
+	userEventDispatcher  *event.DispatchService
+	clusterService       *cluster.ServiceImpl
+	invocationFactory    *cluster.ConnectionInvocationFactory
+	listenerBinder       *cluster.ConnectionListenerBinderImpl
+	smartRouting         bool
+	serviceName          string
+	name                 string
+	logger               ilogger.Logger
+	circuitBreaker       *cb.CircuitBreaker
+}
+
+func newProxy(bundle creationBundle, serviceName string, objectName string) (*proxy, error) {
+	bundle.Check()
+	// TODO: make circuit breaker configurable
+	circuitBreaker := cb.NewCircuitBreaker(
+		cb.MaxRetries(10),
+		cb.MaxFailureCount(10),
+		cb.RetryPolicy(func(attempt int) time.Duration {
+			return time.Duration((attempt+1)*100) * time.Millisecond
+		}))
+	p := &proxy{
+		serviceName:          serviceName,
+		name:                 objectName,
+		requestCh:            bundle.RequestCh,
+		serializationService: bundle.SerializationService,
+		userEventDispatcher:  bundle.UserEventDispatcher,
+		partitionService:     bundle.PartitionService,
+		clusterService:       bundle.ClusterService,
+		invocationFactory:    bundle.InvocationFactory,
+		listenerBinder:       bundle.ListenerBinder,
+		smartRouting:         bundle.SmartRouting,
+		logger:               bundle.Logger,
+		circuitBreaker:       circuitBreaker,
+	}
+	if err := p.create(); err != nil {
+		return nil, err
+	}
+	return p, nil
+}
+
+func (p *proxy) create() error {
+	request := codec.EncodeClientCreateProxyRequest(p.name, p.serviceName)
+	inv := p.invocationFactory.NewInvocationOnRandomTarget(request, nil)
+	p.requestCh <- inv
+	if _, err := inv.Get(); err != nil {
+		return fmt.Errorf("error creating proxy: %w", err)
+	}
+	return nil
+}
+
+func (p *proxy) destroy() error {
```

Comment:
Why is this method kept private? How users are supposed to destroy cluster-side data structures in v4.0?

Reply:
updated